helix-term/src/application.rs

```diff
@@ -822,19 +892,56 @@ impl Application {
         }));
 
         self.event_loop(input_stream).await;
+
         self.close().await?;
         self.restore_term()?;
 
         Ok(self.editor.exit_code)
     }
 
     pub async fn close(&mut self) -> anyhow::Result<()> {
-        self.jobs.finish().await?;
-
-        if self.editor.close_language_servers(None).await.is_err() {
-            log::error!("Timed out waiting for language servers to shutdown");
+        // [NOTE] we intentionally do not return early for errors because we
+        //        want to try to run as much cleanup as we can, regardless of
+        //        errors along the way
+
+        let mut result = match self
+            .jobs
+            .finish(&mut self.editor, &mut self.compositor)
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                log::error!("Error executing job: {}", err);
+                Err(err)
+            }
         };
 
-        Ok(())
+        for doc in self.editor.documents_mut() {
+            if let Some(save_result) = doc.close().await {
+                result = match save_result {
+                    Ok(_) => result,
+                    Err(err) => {
+                        if let Some(path) = doc.path() {
+                            log::error!(
+                                "Error saving document '{}': {}",
+                                path.to_string_lossy(),
+                                err
+                            );
+                        }
+                        Err(err)
+                    }
+                };
+            }
+        }
```

Comment:
Is this awaiting for each item one by one or it is runned asynchronously all of them? I haven't written async/await in some time and I am not sure the behavior here.

Reply:
I believe this will be one by one. This code is only called during the shutdown process though, so this would only matter if the user did something pathological like open a few dozen multigigabyte files, changed them all, and then did a `write-quit-all`.